executor/ddl.go

```diff
@@ -287,3 +290,71 @@ func (e *DDLExec) executeAlterTable(s *ast.AlterTableStmt) error {
 	err := domain.GetDomain(e.ctx).DDL().AlterTable(e.ctx, ti, s.Specs)
 	return errors.Trace(err)
 }
+
+// RestoreTableExec represents a recover table executor.
+// It is built from "admin restore table by job" statement,
+// is used to recover the table that deleted by mistake.
+type RestoreTableExec struct {
+	baseExecutor
+	jobID int64
+}
+
+// Open implements the Executor Open interface.
+func (e *RestoreTableExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Next implements the Executor Open interface.
+func (e *RestoreTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	txn, err := e.ctx.Txn(true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	t := meta.NewMeta(txn)
+	job, err := t.GetHistoryDDLJob(e.jobID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if job == nil {
+		return admin.ErrDDLJobNotFound.GenWithStackByArgs(e.jobID)
+	}
+	if job.Type != model.ActionDropTable {
+		return errors.Errorf("Job %v type is %v, not drop table", job.ID, job.Type)
+	}
+
+	// check gc safe point
+	err = gcutil.ValidateSnapshot(e.ctx, job.StartTS)
```

Comment:
You can move some [code comments](https://github.com/pingcap/tidb/pull/7937/files#diff-71209e5c7ffc793d7bf8779373b48098R186) here.